Bluetooth: host: direction: Add public API to set CTE TX params for adv.

ppryga-nordic opened the pull request:
Add HCI command: HCI_LE_Set_Connectionless_CTE_Transmit_Parameters.
Add public API function to run the command from application level.

ppryga-nordic committed:
Bluetooth: host: Add set connectionless CTE TX params wrapper function

Add implementation of a wrapper for HCI_LE_Set_-
Connectionless_CTE_Transmit_Parameters send command.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: direction: Add public API to set CTE TX params for adv.

Add publicly accessible function to set Constant Tone Extension
parameters for CTE transmission with periodic advertising.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak commented on include/bluetooth/hci.h:
```diff
@@ -1351,6 +1351,27 @@ struct bt_hci_cp_le_set_privacy_mode {
+
+#define BT_HCI_OP_LE_SET_CL_CTE_TX_PARAMS      BT_OP(BT_OGF_LE, 0x0051)
+struct bt_hci_cp_le_set_cl_cte_tx_params {
+	uint16_t handle;
```
```suggestion
	uint8_t handle;
```

ppryga replied:
change applied

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2064,6 +2064,23 @@ static void le_read_tx_power(struct net_buf *buf, struct net_buf **evt)
+				       struct net_buf **evt)
+{
+	struct bt_hci_cp_le_set_cl_cte_tx_params *cmd = (void *)buf->data;
+	uint16_t adv_handle = sys_le16_to_cpu(cmd->handle);
```
```suggestion
```

ppryga replied:
change applied

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2064,6 +2064,23 @@ static void le_read_tx_power(struct net_buf *buf, struct net_buf **evt)
+	uint16_t adv_handle = sys_le16_to_cpu(cmd->handle);
+	uint8_t status;
+
+	status = ll_df_set_cl_cte_tx_params(adv_handle, cmd->cte_len,
```
```suggestion
	status = ll_df_set_cl_cte_tx_params(cmd->handle, cmd->cte_len,
```

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -261,6 +261,11 @@ uint32_t ll_radio_state_is_idle(void);
 /* Direction Finding */
 
+/* Sets CTE transmission parameters for periodic advertising */
+uint8_t ll_df_set_cl_cte_tx_params(uint16_t adv_handle, uint8_t cte_len,
```
```suggestion
uint8_t ll_df_set_cl_cte_tx_params(uint8_t adv_handle, uint8_t cte_len,
```

ppryga replied:
change applied

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -12,6 +12,29 @@
+ *
+ * @return Status of command completion.
+ */
+uint8_t ll_df_set_cl_cte_tx_params(uint16_t adv_handle, uint8_t cte_len,
```
```suggestion
uint8_t ll_df_set_cl_cte_tx_params(uint8_t adv_handle, uint8_t cte_len,
```

ppryga replied:
change applied

ppryga commented:
Requested changes applied.
Also fixed similar issue in hci_df_set_cl_cte_tx_params.

ppryga-nordic committed:
Bluetooth: host: Add DF to initalization sequence

Add Direction Finding to host initialization sequence.
It allows to get information about Direction finding
optional CTE settings supported by controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Because no host code owner has reviewed the PR I've added two things:
- DF initialization (separate commit in PR)
- Checking if DF related features are supported by controller (change commits that provide HCI command wrapper function and public API)

ppryga commented:
@joakimtoe @jhedberg could you do a review?

ppryga commented:
@jhedberg @joerchan @Vudentz @carlescufi @cvinayak could you do a review?

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -7,15 +7,113 @@
 #include "conn_internal.h"
+#include "direction_internal.h"
 
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_CORE)
```
Consider creating your own DEBUG statement, HCI_CORE enables a lot

ppryga replied:
I remove it now because it is not required now. I will add separate DF debug entry in Kconfig when  such requirement arrives.

joerchan replied:
Why is it not required? Without this line it will be enabled always due to this:
```
/* common/log.h */
#if !defined(BT_DBG_ENABLED)
#define BT_DBG_ENABLED 1
#endif
```

ppryga replied:
Sorry, I didn't spot that. I'll fix it.

ppryga replied:
Code is corrected and new Kconfig `CONFIG_BT_DEBUG_DF` added.

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -7,15 +7,113 @@
 
+/* @brief Antenna information for LE Direction Finding */
+struct bt_le_df_ant_info {
+	/* Bitfiled holding optional switching and sampling rates */
```
typo: Bitfield

ppryga replied:
corrected

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -131,3 +229,63 @@ static int hci_df_set_conn_cte_tx_param(struct bt_conn *conn, uint8_t cte_types,
+	df_ant_info.max_cte_len = max_cte_len;
+	df_ant_info.num_ant = num_ant;
+
+	BT_INFO("DF initialized.");
```
BT_DBG

ppryga replied:
changed as requested

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -7,15 +7,113 @@
+		}
+		__ASSERT_NO_MSG((sizeof(*cp) + params->num_ant_ids) <
+				UINT8_MAX);
+	}
```
cte_type not set to a valid value would be passed to the controller now.

ppryga replied:
added missing verification of cte_type correctness

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -131,3 +229,63 @@ static int hci_df_set_conn_cte_tx_param(struct bt_conn *conn, uint8_t cte_types,
+		return -EINVAL;
+	}
+
+	if (atomic_test_bit(adv->flags, BT_PER_ADV_CTE_ENABLED)) {
```
I cannot see that this flag is set anywhere.

ppryga replied:
The flag would be set in function that handles CTE enable HCI command. It will be delivered in following PR.

joerchan replied:
@ppryga  The check of this flag should be moved to that commit as well then.

ppryga replied:
Test of flag removed.

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -131,3 +229,63 @@ static int hci_df_set_conn_cte_tx_param(struct bt_conn *conn, uint8_t cte_types,
+		return err;
+	}
+
+	atomic_set_bit(adv->flags, BT_PER_ADV_CTE_PARAMS_SET);
```
This is never read, do you need it?

If it was supposed to be CTE_ENABLED I would suggest to set it using state_set mechanism so that it is set during the handling of the command complete event

ppryga replied:
It is needed for function that will handle CTE enable command.

joerchan replied:
If this will be used in a future addition then it should be moved to that commit so that the commits are self-contained.

ppryga replied:
@joerchan Isn't it self-contained? I assume the commit shoud provide complete implementation of some functionality. In this case it is implementation of `bt_le_df_set_adv_cte_tx_params`. This is part of implementation of the function.
If I provide it as part of implementation of `bt_df_set_adv_cte_tx_enabled` then one may ask: "why implementation of enable function includes changes in other functions?". 
I thought the goal is to provide complete implementation if that is possible. 

Am I wrong here?

joerchan replied:
Because they don't do anything at the moment. You should add the flags when you use them. 
Self-contained as a feature, not a function.

ppryga replied:
Well, I've added them now, because for me it was part of complete feature - to set up CTE parameters (sorry for being unclear). I can remove it if you whish that and add it in following PR.
I just want to get feeling what is expected to be delivered in PRs:
- complete feature as it is required by protocol specification, this means for me add these flags here because they are part of specification and can be added.
-  complete feature in the meaning of code that is actually in use? So not all things from protocol specification are be implemented but no unused code is present.

joerchan replied:
@ppryga I'm mostly referring to this from the contributing guidelines:
```
Each commit must address a single identifiable issue and must be logically self-contained. Unrelated changes should be submitted as separate commits.
```

The reason why I wanted them to be grouped together is that when you look at just the single commit it looks really strange, it looks like something has been forgotten and a possible bug
And when the follow-up would have been commit in a PR later there could be hundreds of commits in between.
Someone looking through the history, either debugging or curious might, might be confused and think the same as I did.

I think the easiest would have been to add `bt_df_set_adv_cte_tx_enabled` in the same commit as you added `bt_df_set_adv_cte_tx_param`. That way you could  look at the all together.

Sorry if I'm making too big a deal of out it.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+ * The structure holds information required to setup CTE transmission
+ * in periodic advertising.
+ */
+struct bt_le_df_adv_cte_tx_params {
```
I think we should not add `_le_` as infix here, in order to get shorter API names since this is completely new.

Also the other places in bluetooth.h we use `param` instead of `params` so should align there (I noticed gatt.h uses it different, which is a shame, but this should be aligned with bluetooth.h IMO).

ppryga replied:
infix removed here and in `bt_df_set_adv_cte_tx_params` function name.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+	uint8_t  *ant_ids;
+};
+
+/**
```
Please start the @brief on the first line

ppryga replied:
corrected

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -7,15 +7,113 @@
+#define DF_AOA_RX_1US_SUPPORT(supp)             (DF_SUPP_TEST(supp, \
+						BT_HCI_LE_1US_AOA_RX))
+
+/* @brief Function sets CTE parameters for connectionless transmission
```
You don't have to document internal functions, atleast not with doxygen.

When documenting internally it is more important to document for the developers that maintain this.
In my opinion this documentation does not provide much information since the information is already contained in the function name, the input argument name and the return type.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+ * The Constant Tone Extension parameters can only be set or updated on an
+ * extended advertisment set which has already set periodic adverising
+ * parameters. Update of Constant Tone Extension parameters may not be attempted
+ * while Constnat Tone Extension is enabled on particular extended advertisement
```
Typo: Constant Tone Extension

ppryga replied:
corrected

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+ *
+ * The Constant Tone Extension parameters can only be set or updated on an
+ * extended advertisment set which has already set periodic adverising
+ * parameters. Update of Constant Tone Extension parameters may not be attempted
```
The updated sentence is unclear what means here now.

Do I need to stop both the periodic advertiser (bt_le_per_adv_stop) and the extended advertiser (bt_le_ext_adv_stop) in order to be able to update the CTE parameters?

Suggest to reword "attempted". You could always attempt, but would return an error code. Should use "allowed" instead.

ppryga replied:
I've corrected the description.
First part of it was my wrong understanding of specification. Parameters may be set on existing advertising set.
That is obvious because one has to provide bt_le_ext_adv reference.
Second part is related with function that will be provided in following PR. 
The function enables transmission of CTE with periodic advertising.
To update CTE parameters, transmission must be disabled before.

I've updated the descritpion.

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -7,15 +7,113 @@
+		    !params->ant_ids) {
+			return -EINVAL;
+		}
+		__ASSERT_NO_MSG((sizeof(*cp) + params->num_ant_ids) <
```
Instead of asserting we could check the valid range.
0x02 to 0x4B. that should cover the overflow.

ppryga replied:
Valid range is actually verified few lines above. Because of that assert was redundant and removed.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+ * in periodic advertising.
+ */
+struct bt_le_df_adv_cte_tx_params {
+	/* Length of CTE in 8us units */
```
Need to use `/**` for doxygen. (rest of this struct)

ppryga replied:
Corrected.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,47 @@
+ * set.
+ *
+ * @param[in] adv           Advertising set object.
+ * @param[in] params         Constant Tone Extension parameters.
```
The alignment is off here.

ppryga replied:
Corrected.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -0,0 +1,45 @@
+ * advertising set.
+ *
+ * Update of Constant Tone Extension parameters is not allowed while
+ * transmission of Constant Tone Extension is enabled on particular extended
```
Is the rule about enabling CTE transmission also part of the follow up PR?
I suggest we move this text block until we have added that API so that we can review it in its complete form.

ppryga replied:
The rule is actually related with Core spec: Vol 4, Part E 7.8.80 LE Set Connectionless CTE Transmit Parameters command:
`If the Host issues this command when Constant Tone Extensions have been
enabled in the advertising set, the Controller shall return the error code
Command Disallowed (0x0C).`

joerchan replied:
Right, but need to explain what that means in terms of the zephyr Host API.
When is CTE enabled in the advertising set? periodic_adv start with CTE configured?

ppryga replied:
CTE is enabled by call to another function, the one that will be provided in following PR. 
The function here is responsible for setting CTE parameters and related with HCI command `BT_HCI_OP_LE_SET_CL_CTE_TX_PARAMS`.
The CTE enable is another API call that will be named: `bt_df_set_adv_cte_tx_enabled`. This function is related with `BT_HCI_OP_LE_SET_CL_CTE_TX_ENABLE` HCI command.

ppryga commented:
@jhedberg @Vudentz could you do a review?  Your approval is missing.
